File: javascript/src/Structural/Composite/TinOfTeaBags.js
(function() {
  var __hasProp = {}.hasOwnProperty,
    __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; };

  define(['Structural/Composite/TeaBags'], function(TeaBags) {
    var TinOfTeaBags;
    TinOfTeaBags = (function(_super) {
      __extends(TinOfTeaBags, _super);

      function TinOfTeaBags(name) {
        this.name = name;
        this.teaBagList = [];
      }

      TinOfTeaBags.prototype.countTeaBags = function() {
        var list, teaBags, totalTeaBags, _i, _len;
        totalTeaBags = 0;
        list = this.createListIterator();
        for (_i = 0, _len = list.length; _i < _len; _i++) {
          teaBags = list[_i];
          totalTeaBags += teaBags.countTeaBags();
        }
        return totalTeaBags;
      };

      TinOfTeaBags.prototype.add = function(teaBagsToAdd) {
        teaBagsToAdd.setParent(this);
        return this.teaBagList.push(teaBagsToAdd);
      };

      TinOfTeaBags.prototype.remove = function(teaBagsToRemove) {
        var list, list_size, teaBag;
        list = this.createListIterator();
        list_size = list.length;
        this.teaBagList = (function() {
          var _i, _len, _results;
          _results = [];
          for (_i = 0, _len = list.length; _i < _len; _i++) {
            teaBag = list[_i];
            if (teaBag !== teaBagsToRemove) {
              _results.push(teaBag);
            }
          }
          return _results;
        })();
        return list_size !== this.teaBagList.length;
      };

      TinOfTeaBags.prototype.createListIterator = function() {
        return this.teaBagList;
      };

      return TinOfTeaBags;

    })(TeaBags);
    return TinOfTeaBags;
  });

}).call(this);

File: go/structural/composite/TinOfTeaBags.go
package composite


//TinOfTeaBags - one composite extension - the "node"

type TinOfTeaBags struct {
	Teabags *TeaBagsAbstract
}

func NewTinOfTeaBags( nameIn string ) *TinOfTeaBags {  
	obj := new(TinOfTeaBags)
	obj.Teabags = new(TeaBagsAbstract)
	obj.Teabags.SetName(nameIn)
	obj.Teabags.TeabagsList = make([]TeaBagsInterface,0)
	return obj
}
   
func (this *TinOfTeaBags) CountTeaBags() int {
	totalTeaBags := 0
	if this.Teabags.TeabagsList != nil {
		for _,v := range this.Teabags.TeabagsList {
			totalTeaBags += v.CountTeaBags()
		}
	}
	return totalTeaBags
}
   
func (this *TinOfTeaBags) Add( tb TeaBagsInterface) bool {
	//tb.SetParent( this)
	this.Teabags.TeabagsList = append( this.Teabags.TeabagsList, tb)
	return true
}
   
func (this *TinOfTeaBags) Remove( tb TeaBagsInterface) bool {
   foundI := -1
   for i, v := range this.Teabags.TeabagsList {
      if tb == v {
         foundI = i
         break;
      }
   }
   if foundI != -1 {
      this.Teabags.TeabagsList[foundI] = this.Teabags.TeabagsList[len(this.Teabags.TeabagsList)-1]
      this.Teabags.TeabagsList = this.Teabags.TeabagsList[:len(this.Teabags.TeabagsList)-1]
      return true
   }
   return false
}

func (this *TinOfTeaBags) SetParent(parentIn *TeaBagsInterface ) {
   this.Teabags.SetParent(parentIn)
}

func (this *TinOfTeaBags) GetParent() *TeaBagsInterface {
   return this.Teabags.GetParent()
}

func (this *TinOfTeaBags) SetName(nameIn string) {
   this.Teabags.SetName(nameIn )
}

func (this *TinOfTeaBags) GetName() string {
   return this.Teabags.GetName()
}


